tutorials/04-hidden-markov-model/index.qmd

```diff
@@ -190,4 +195,85 @@ stationary. We can use the diagnostic functions provided by [MCMCChains](https:/
 heideldiag(MCMCChains.group(chn, :T))[1]
 ```
 
-The p-values on the test suggest that we cannot reject the hypothesis that the observed sequence comes from a stationary distribution, so we can be reasonably confident that our transition matrix has converged to something reasonable.
\ No newline at end of file
+The p-values on the test suggest that we cannot reject the hypothesis that the observed sequence comes from a stationary distribution, so we can be reasonably confident that our transition matrix has converged to something reasonable.
+
+## Efficient Inference With The Forward Algorithm
+
+While the above method works well for the simple example in this tutorial, some users may desire a more efficient method, especially when their model is more complicated. One simple way to improve inference is to marginalize out the hidden states of the model with an appropriate algorithm, calculating only the posterior over the continuous random variables. Not only does this allow more efficient inference via Rao-Blackwellization, but now we can sample our model with `NUTS()` alone, which is usually a much more performant MCMC kernel.
+
+Thankfully, [HiddenMarkovModels.jl](https://github.com/gdalle/HiddenMarkovModels.jl) provides an extremely efficient implementation of many algorithms related to Hidden Markov Models. This allows us to re-write our model as:
+
+```{julia}
+#| output: false
+using HiddenMarkovModels
+using FillArrays
+using LinearAlgebra
+using LogExpFunctions
+
+
+@model function BayesHmm2(y, K)
+    m ~ Bijectors.ordered(MvNormal([1.0, 2.0, 3.0], 0.5I))
+    T ~ filldist(Dirichlet(Fill(1/K, K)), K)
+
+    hmm = HMM(softmax(ones(K)), copy(T'), [Normal(m[i], 0.1) for i in 1:K])
+    Turing.@addlogprob! logdensityof(hmm, y)
+end
+
+chn2 = sample(BayesHmm2(y, 3), NUTS(), 1000)
+```
+
+
+We can compare the chains of these two models, confirming the posterior estimate is similar (modulo label switching concerns with the Gibbs model):
+```{julia}
+#| code-fold: true
+#| code-summary: "Plotting Chains"
+
+plot(chn["m[1]"], label = "m[1], Model 1, Gibbs", color = :lightblue)
+plot!(chn2["m[1]"], label = "m[1], Model 2, NUTS", color = :blue)
+plot!(chn["m[2]"], label = "m[2], Model 1, Gibbs", color = :pink)
+plot!(chn2["m[2]"], label = "m[2], Model 2, NUTS", color = :red)
+plot!(chn["m[3]"], label = "m[3], Model 1, Gibbs", color = :yellow)
+plot!(chn2["m[3]"], label = "m[3], Model 2, NUTS", color = :orange)
+```
+
+
+### Recovering Marginalized Trajectories
+
+We can use the `viterbi()` algorithm, also from the `HiddenMarkovModels` package, to recover the most probable state for each parameter set in our posterior sample:
+```{julia}
+#| output: false
+@model function BayesHmmRecover(y, K, IncludeGenerated = false)

```

Comment:
As above – if we do `setprogress!(false)` above can we enable the output as usual?

Reply:
I think we could, but do we want to? I assume most users will want the progress updates in their REPL when working through the tutorial, but they just aren't a good fit for the published quarto format. 

As long as we are not actually recommending users use `setprogress!(false)` (which I think we aren't?), it seems appropriate to just disable this in the chunk options for the rendered docs.

Reply:
In fact (I totally missed this earlier) but there is already a call to `setprogress!(false)` earlier in the page (which is hidden). Many other pages do call `setprogress!(false)` so it wouldn't be too out of place here, so I feel like it should be alright to just remove the `output: false` lines. I'd be happy to add a clarification about this on the get started page in a separate PR.

Reply:
(By the way, I'm assuming that the output you wanted to suppress was indeed the progress logging. Do correct me if that's not the case)